CONTRIBUTING.md

```diff
@@ -39,6 +39,10 @@ If you want to debug this WAR file without using Maven plugins,
 You can just start the executable with [Remote Debug Flags]
 and then attach IDE Debugger to it.
 
+To compile and launch a development instance, run:
+
+    mvn hudson-dev:run
```

Comment:
```suggestion
    mvn jenkins-dev:run
```

Also needs to be in the `war/` directory (I think? At least that's what I'm doing…).

This does not explain what the difference is to the paragraphs above it (and exactly what should be mentioned, while JENKINS-55150 is unresolved?)

Reply:
Running `mvn jenkins-dev:run` from the `war` directory does not work for me.

![Screenshot](https://user-images.githubusercontent.com/1390036/67926270-df913a80-fbb5-11e9-94a9-01adc599ef22.png)

Starting the `jenkins.war` as above [1] (without any additional parameters), launches the installations wizard requires a install into `JENKINS_HOME`.  Using `mvn hudson-dev:run` from the source root starts the instance directly without installation (no login).

```
> git pull origin master
From https://github.com/jenkinsci/jenkins
 * branch                  master     -> FETCH_HEAD
Already up to date.
> java -version
java version "1.8.0_231"
Java(TM) SE Runtime Environment (build 1.8.0_231-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.231-b11, mixed mode)
```

[1]
```
> java -jar war\target\jenkins.war
Running from: D:\xxx\CI\jenkins\war\target\jenkins.war
webroot: $user.home/.jenkins
2019-10-31 07:21:28.826+0000 [id=1]     INFO    org.eclipse.jetty.util.log.Log#initialized: Logging initialized @897ms to org.eclipse.jetty.util.log.JavaUtilLog
2019-10-31 07:21:28.950+0000 [id=1]     INFO    winstone.Logger#logInternal: Beginning extraction from war file
2019-10-31 07:21:31.017+0000 [id=1]     WARNING o.e.j.s.handler.ContextHandler#setContextPath: Empty contextPath
2019-10-31 07:21:32.054+0000 [id=1]     INFO    org.eclipse.jetty.server.Server#doStart: jetty-9.4.z-SNAPSHOT; built: 2019-05-02T00:04:53.875Z; git: e1bc35120a6617ee3df052294e433f3a25ce7097; jvm 1.8.0_231-b11
2019-10-31 07:21:32.824+0000 [id=1]     INFO    o.e.j.w.StandardDescriptorProcessor#visitServlet: NO JSP Support for /, did not find org.eclipse.jetty.jsp.JettyJspServlet
2019-10-31 07:21:32.911+0000 [id=1]     INFO    o.e.j.s.s.DefaultSessionIdManager#doStart: DefaultSessionIdManager workerName=node0
2019-10-31 07:21:32.918+0000 [id=1]     INFO    o.e.j.s.s.DefaultSessionIdManager#doStart: No SessionScavenger set, using defaults
2019-10-31 07:21:32.927+0000 [id=1]     INFO    o.e.j.server.session.HouseKeeper#startScavenging: node0 Scavenging every 600000ms
2019-10-31 07:21:33.593+0000 [id=1]     INFO    hudson.WebAppMain#contextInitialized: Jenkins home directory: C:\Users\t.sparber\.jenkins found at: $user.home/.jenkins
2019-10-31 07:21:33.819+0000 [id=1]     INFO    o.e.j.s.handler.ContextHandler#doStart: Started w.@7caa550{Jenkins v2.203-SNAPSHOT,/,file:///C:/Users/XX/.jenkins/war/,AVAILABLE}{C:\Users\XX\.jenkins\war}
2019-10-31 07:21:33.881+0000 [id=1]     INFO    o.e.j.server.AbstractConnector#doStart: Started ServerConnector@221a3fa4{HTTP/1.1,[http/1.1]}{0.0.0.0:8080}
2019-10-31 07:21:33.882+0000 [id=1]     INFO    org.eclipse.jetty.server.Server#doStart: Started @5954ms
2019-10-31 07:21:33.891+0000 [id=23]    INFO    winstone.Logger#logInternal: Winstone Servlet Engine v4.0 running: controlPort=disabled
2019-10-31 07:21:35.252+0000 [id=30]    INFO    jenkins.InitReactorRunner$1#onAttained: Started initialization
2019-10-31 07:21:35.292+0000 [id=33]    INFO    jenkins.InitReactorRunner$1#onAttained: Listed all plugins
2019-10-31 07:21:37.556+0000 [id=41]    INFO    jenkins.InitReactorRunner$1#onAttained: Prepared all plugins
2019-10-31 07:21:37.565+0000 [id=41]    INFO    jenkins.InitReactorRunner$1#onAttained: Started all plugins
2019-10-31 07:21:37.577+0000 [id=41]    INFO    jenkins.InitReactorRunner$1#onAttained: Augmented all extensions
2019-10-31 07:21:38.701+0000 [id=29]    INFO    jenkins.InitReactorRunner$1#onAttained: Loaded all jobs
2019-10-31 07:21:38.724+0000 [id=56]    INFO    hudson.model.AsyncPeriodicWork#lambda$doRun$0: Started Download metadata
2019-10-31 07:21:38.749+0000 [id=56]    INFO    hudson.util.Retrier#start: Attempt #1 to do the action check updates server
2019-10-31 07:21:39.908+0000 [id=31]    INFO    o.s.c.s.AbstractApplicationContext#prepareRefresh: Refreshing org.springframework.web.context.support.StaticWebApplicationContext@5dda3db7: display name [Root WebApplicationContext]; startup date [Thu Oct 31 08:21:39 CET 2019]; root of context hierarchy
2019-10-31 07:21:39.908+0000 [id=31]    INFO    o.s.c.s.AbstractApplicationContext#obtainFreshBeanFactory: Bean factory for application context [org.springframework.web.context.support.StaticWebApplicationContext@5dda3db7]: org.springframework.beans.factory.support.DefaultListableBeanFactory@6ffb644
2019-10-31 07:21:39.927+0000 [id=31]    INFO    o.s.b.f.s.DefaultListableBeanFactory#preInstantiateSingletons: Pre-instantiating singletons in org.springframework.beans.factory.support.DefaultListableBeanFactory@6ffb644: defining beans [authenticationManager]; root of factory hierarchy
2019-10-31 07:21:40.201+0000 [id=31]    INFO    o.s.c.s.AbstractApplicationContext#prepareRefresh: Refreshing org.springframework.web.context.support.StaticWebApplicationContext@75ea4627: display name [Root WebApplicationContext]; startup date [Thu Oct 31 08:21:40 CET 2019]; root of context hierarchy
2019-10-31 07:21:40.201+0000 [id=31]    INFO    o.s.c.s.AbstractApplicationContext#obtainFreshBeanFactory: Bean factory for application context [org.springframework.web.context.support.StaticWebApplicationContext@75ea4627]: org.springframework.beans.factory.support.DefaultListableBeanFactory@15d9bb81
2019-10-31 07:21:40.205+0000 [id=31]    INFO    o.s.b.f.s.DefaultListableBeanFactory#preInstantiateSingletons: Pre-instantiating singletons in org.springframework.beans.factory.support.DefaultListableBeanFactory@15d9bb81: defining beans [filter,legacy]; root of factory hierarchy
2019-10-31 07:21:40.440+0000 [id=31]    INFO    jenkins.install.SetupWizard#init:

*************************************************************
*************************************************************
*************************************************************

Jenkins initial setup is required. An admin user has been created and a password generated.
Please use the following password to proceed to installation:

XXXX

This may also be found at: C:\Users\XXXX\.jenkins\secrets\initialAdminPassword

*************************************************************
*************************************************************
```

Reply:
As the `mvn jenkins-dev:run` is the documented command, there might be something wrong with my setup. It is a clean checkout from this GitHub repo. Is there anything additonal that needs to be done?

Reply:
I need to `mvn install` first from the top-level directory, then can `mvn jenkins-dev:run` from the `war` directory (but the error is a different one).

Perhaps it's due to the following in my `.m2/settings.xml`?

```
  <pluginGroups>
    <pluginGroup>org.jenkins-ci.tools</pluginGroup>
  </pluginGroups>
```

That said, there's a full command line that would work, see my workaround comment (minus the explicit version?) in JENKINS-55150.

Reply:
Don't know. https://github.com/jenkinsci/jenkins/blob/b9cda5854afd310aea4acec111fa942e8546bcd6/pom.xml#L507-L515 looks like it should enable the former, not the latter.

FWIW I always got an empty directory served up when I ran `jenkins-dev:run` from the top level of the project -- could you confirm an actual Jenkins is running as expected when run from the top level?

Reply:
Running `mvn org.jenkins-ci.tools:maven-jenkins-dev-plugin:run` also works without the additional changes to the `.m2/settings.xml` from the top-level.

Digging deeper I found this
https://issues.jenkins-ci.org/browse/JENKINS-23364
and
https://github.com/jenkinsci/maven-hudson-dev-plugin/pull/1
and
https://groups.google.com/forum/#!topic/jenkinsci-dev/UNW6bWjcgF4

So the behaviour is "known", but quite unexpected for a new user (new to Jenkins and Maven).
I also just saw that it is documented here, that changes to the settings.xml are needed:
https://wiki.jenkins.io/display/JENKINS/Building+Jenkins#BuildingJenkins-DebuggingJenkins
(This documentation also not easily found)

So is see following options for this PR to update `CONTRIBUTING.md`:
1. Recommend the `mvn hudson-dev:run` as suggested
2. Recommend `mvn org.jenkins-ci.tools:maven-jenkins-dev-plugin:run`
3. Recommend `mvn jenkins-dev:run` and state again the needed changes to the `.m2/settings.xml`
4. Add a Link to https://wiki.jenkins.io/display/JENKINS/Building+Jenkins#BuildingJenkins-DebuggingJenkins (and highlight the changes needed in `.m2/settings.xml`)

What do you prefer?

Reply:
IMO: For a "quick start" type doc, 4 seems too indirect/wordy. 3 seems unnecessarily complex for the little benefit it brings. 2 is annoying to type. 1 has the obsolete name, but looks like the least evil. Perhaps add an HTML comment to the doc, to explain the tradeoffs involved, so it's clear in the future why this one was chosen, in case someone tries to be helpful and `s/hudson/jenkins/` it.

Users involved in developing Jenkins quickly need to become used to the numerous internal Hudson references anyway.

Thanks for digging into this!

Reply:
added a comment with some reasoning.

Reply:
FWIW regarding the "top level directory" problem, it seems this was fixed since `org.jenkins-ci.tools:maven-jenkins-dev-plugin:9.4.5.v20170502-jenkins-1`, the version I'm still using due to JENKINS-55150.